### ISSUE TITLE
recommended solution for multiple MQs

### ISSUE THREAD

**Issue by VinSpee** [ReactTraining/react-media#38]
Hi,

what's the current solution for multiple MQs? Multiple media components? How should I compose something like that?

A solution like this _feels_ right to me:

```js
      <Media
        query={{
          sm: {
            minWidth: theme.sizes.z5,
          },
          md: {
            minWidth: theme.sizes.z10,
          },
        }}
      >
        {({ sm, md }) => (
           <div>
               { sm && <span>I am small</span> }
               { md && <span> I am medium</span> }
           </div>
        );
        }
      </Media>
```

at small would render as:

```html
<div>
   <span>I am small</span>
</div>
```

and at medium:

```html
<div>
    <span>I am small</span>
    <span>I am medium</span>
</div>
```

Thoughts?

**Comment by mjackson** [comment]
To avoid overloading the `query` prop, I think we should have a `queries` prop that expects the kind of object you're describing, and it would use the keys as you've shown. Care to make a PR?

**Comment by VinSpee** [comment]
Sounds like a good idea. I'll give it a whirl!

**Comment by VinSpee** [comment]
@mjackson 

here's the outline I'm thinking:

`propTypes` change to:

```
  static propTypes = {
    query: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.object,
      PropTypes.arrayOf(PropTypes.object.isRequired)
    ]),
    queries: PropTypes.shape({
      [PropTypes.string]: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.object,
        PropTypes.arrayOf(PropTypes.object.isRequired)
      ]),
    }),
    render: PropTypes.func,
    children: PropTypes.oneOfType([
      PropTypes.node,
      PropTypes.func
    ])
  }
```

now in `componentWillMount` we need to check for either prop, if `query`, keep functionality the same. If `queries`

- create an array of `json2mq` strings, 
- set `this.mediaQueryList` to be an array `matchMedia` objects for each of the queries
- add a listener to each of them that runs `this.updateMatches`

then, in `updateMatches`

- if `query` prop, keep functionality the same
- if `queries` prop, set state to an object where the key is the name of the query, the value is `true` / `false` according to the `matchMedia.matches` value.

sound sound?

**Comment by VinSpee** [comment]
for the curious: here's a webpackbin demo to play around with

http://www.webpackbin.com/E1Ce0d0KG

**Comment by ryanflorence** [comment]
I've had a version of this a couple years ago that had this api:

`<Media queries={{ small: '...', large: '...' }}/>`

**Comment by VinSpee** [comment]
@ryanflorence here's the API I was suggesting, the same as your idea:

https://github.com/ReactTraining/react-media/pull/39/commits/a0665e881278ba83280fca08229e7af5bec14336#diff-02d7a6fb8ca14af07e3d2ce4f14dea77R77

```js
<Media queries={{
     small: (this value could be anything that a `query` can be currently),
     large: ...same,
/>

**Comment by stephenjwatkins** [comment]
I've been wanting this as well. I hope this change makes it in!

**Comment by mjackson** [comment]
I know it's a little late for discussion around this (especially since I just merged #39 this morning) but what's to prevent you from doing this @VinSpee?

```jsx
<div>
  <Media query={{ minWidth: theme.sizes.z5 }} render={() => <span>I am small</span>}/>
  <Media query={{ minWidth: theme.sizes.z10 }} render={() => <span>I am medium</span>}/>
</div>
```

You'd get the same result, right?

Only reason I'm wondering is because the code is a lot more complex now just to support the `queries` prop, but it seems like you could achieve the same result with more than one `<Media>` element.

**Comment by mjackson** [comment]
Ping @souporserious who I know was also thinking about this problem

**Comment by VinSpee** [comment]
in the specific example I provided, that would be a fine solution, but if I'm doing something like changing classnames based on breakpoint, it becomes very very messy quickly

**Comment by souporserious** [comment]
Same thoughts as @VinSpee, the ability to pass multiple queries can help with overall responsive design. With the latest addition I can now build a context provider and HoC around `Media` and pass my breakpoints down to child components really easy. Similar to [this project](https://github.com/jxnblk/react-media-context).

**Comment by mjackson** [comment]
Thanks for the feedback, @VinSpee and @souporserious. It'd be great to have a test that shows off what you're talking about, just so I can easily remember why the added complexity is justified. Maybe something like this?

```jsx
  <Media queries={{ small: '...', large: '...' }} render={({ small, large }) => (
    <div className={small ? 'small' : large ? 'large' : 'medium'}/>
  )}/>
```

**Comment by VinSpee** [comment]
happy to add a test for that case. incoming.

**Comment by mjackson** [comment]
@VinSpee Maybe just modify the existing test you wrote for this feature.

**Comment by gribnoysup** [comment]
Was waiting for this PR for exactly the same reason, it is not always about rendering/not rendering component on some breakpoint, but changing its props to. Great work @VinSpee!

**Comment by mjackson** [comment]
I'm starting to think it may be better to just have a separate component for this behavior, instead of adding a `<Media queries>` prop. There's no reason 2 completely separate code paths need to live in the same component. But I'll probably do that refactoring on my own.

Thanks again for your work here, @VinSpee! 💃

**Comment by souporserious** [comment]
What are your thoughts for the other component @mjackson? I've been playing around with this and had some ideas.

A way to wrap a tree and provide common breakpoints to all components, would most likely just wrap your highest level component:
`<MediaQueryProvider queries={{ sm: {  maxWidth: 300 }, md: { minWidth: 301 }  }}>`

HoC to easily grab the currently matched queries from MediaQueryProvider:
`withMatchedQueries()`

And finally, you can use the `Media` component like before, _but_ I think it would be cool if you got access to the queries if you're using the `MediaQueryProvider` component so you can do something like:
`<Media query="sm">` or just pass a regular query `<Media query={{ minWidth: 300, maxWidth: 500 }}>`

Thoughts?

**Comment by mjackson** [comment]
I'm a big fan of just using straight components instead of providers + HOCs. I was just thinking it would be great to have support for multiple queries as a completely separate component instead of a different prop to the same component. Right now, all of the code in `<Media>` essentially switches like this:

```js
if (query) {
  // do something for a single media query
} else if (queries) {
  // do something for multiple media queries
}
```

Thing is, most people are only actually going to use either `query` or `queries`, not both. Also, we don't currently gracefully handle the case where someone uses **both** props (i.e. `<Media query queries>`), so that's a little weird. I know I'm the one who proposed `<Media queries>`, but in hindsight I think it wasn't a good idea.

Seems like it could be handled much more easily with a completely separate component:

```jsx
<Media query>
<Multi queries>
```

@souporserious You can get the convenience of declaring your media query only once by just re-using your media query object:

```jsx
const mediaQueries = {
  sm: ...
}

<Media query={mediaQueries.sm}/>
<Media query={mediaQueries.sm}/>
```

One big upside to this approach is you don't have to use strings to refer to your media queries and then go back up to the provider to see what they mean.

**Comment by VinSpee** [comment]
I'm not opposed to a multiple queries export save for the weirdness it creates around importing (feels weird to have one as the `default` and one named). maybe a v2 after all?

**Comment by mjackson** [comment]
@VinSpee Agree. If we do go with 2 components, I think a v2 makes sense.

**Comment by souporserious** [comment]
Love it! 😄  it's so easy to make things more complicated when they can actually be more simple and explicit 😆

**Comment by mjackson** [comment]
This issue has me thinking about another thing: the `render` prop. With a single media query, it makes a lot of sense to have a `render` prop that doesn't render unless the query matches. But with multiple queries, it's less clear what that prop is for.

**Comment by nezed** [comment]
@mjackson I'm voting for reopen this issue cause merged #39 #58 are dropped from master.
We still need `queries` prop

**Comment by mjackson** [comment]
@nezed Good call. I opened #69 instead. Let's follow up there.
